sopel/plugins/rules.py

```diff
@@ -0,0 +1,1189 @@
+        :return: ``True`` if the rule exists, ``False`` otherwise
+
+        The optional parameter ``plugin`` can be provided to limit the rules
+        to the ones of said plugin.
+        """
+        rules = (
+            itertools.chain(*self._rules.values())
+            if plugin is None
+            else self._rules.get(plugin, [])
+        )
+        return any(rule.get_rule_label() == label for rule in rules)
+
+    def has_command(self, name, follow_alias=False, plugin=None):
+        """Tell if the manager knows a command with this ``name``.
+
+        :param str label: Label of the rule to look for
+        :param bool follow_alias: Optional filter to follow alias or not
+        :param str plugin: Optional filter on the plugin name
+        :return: ``True`` if the command exists, ``False`` otherwise
+
+        By default, this method doesn't search commands by their aliases. If
+        the optional parameter ``follow_alias`` is ``True``, then it will
+        detect commands by their aliases too::
+
+            >>> command = Command('hi', prefix='"', aliases=['hey'])
+            >>> manager.register_command(command)
+            >>> manager.has_command('hi')
+            True
+            >>> manager.has_command('hey')
+            False
+            >>> manager.has_command('hey', follow_alias=True)
+            True
+
+        The optional parameter ``plugin`` can be provided to limit the commands
+        to the ones of said plugin.
+        """
+        return _has_named_rule(self._commands, name, follow_alias, plugin)
+
+    def has_nick_command(self, name, follow_alias=False, plugin=None):
+        """Tell if the manager knows a nick command with this ``name``.
+
+        :param str label: Label of the rule to look for
+        :param bool follow_alias: Optional filter to follow alias or not
+        :param str plugin: Optional filter on the plugin name
+        :return: ``True`` if the command exists, ``False`` otherwise
+
+        This method works like :meth:`has_command`, but with nick commands.
+        """
+        return _has_named_rule(self._nick_commands, name, follow_alias, plugin)
+
+    def has_action_command(self, name, follow_alias=False, plugin=None):
+        """Tell if the manager knows an action command with this ``name``.
+
+        :param str label: Label of the rule to look for
+        :param bool follow_alias: Optional filter to follow alias or not
+        :param str plugin: Optional filter on the plugin name
+        :return: ``True`` if the command exists, ``False`` otherwise
+
+        This method works like :meth:`has_command`, but with action commands.
+        """
+        return _has_named_rule(
+            self._action_commands, name, follow_alias, plugin)
+
+    def get_all_commands(self):
+        """Retrieve all the register commands, by plugin."""
+        # expose a copy of the registered commands
+        return self._commands.items()
+
+    def get_all_nick_commands(self):
```

Comment:
These methods reflect the need to generate the `command_groups` and `doc` attributes. Because these don't care about anonymous rules and action commands, I didn't implement methods for them: the focus of this PR is to do the bare minimum to be backward compatible, and to allow new features.

So: valid point, and we need to address that in a future PR.

Reply:
By the way, note how short the docstrings are for these methods. I really didn't want to invest too much time in something I wasn't 100% sure about.

Reply:
> Because these don't care about anonymous rules and action commands, I didn't implement methods for them

Arguably, action commands *should* be included in command groups/docs. Even if it's not used immediately, I'd implement the relevant method here. You certainly aren't afraid of planning for the future (see: `has_*` methods). 😛

At this point, I should mention that I think rules should have a "fetch all" too, since we do have #311 (now sopel-irc/sopel.chat#23) to worry about later.

Reply:
Yeah, let's do that in a future PR. I'd like to improve upon what we have here, without having to re-review 4000 lines of change.